Improving possible confusing instructions

tixel opened the pull request:
The instruction tell the reader to search and replace USERNAME in 2 scripts. But some USERNAME strings should not be replaced. Changed the string that need to be replaced with YOUR_USERNAME and changed the instructions to match that.

tixel committed:
Improving possible confusing instructions

The instruction tell the reader to search and replace USERNAME in 2 scripts. But some USERNAME strings should not be replaced.
Changed the string that need to be replaced with YOUR_USERNAME and changed the instructions to match that.

Signed-off-by: Tim <timbosschaerts@gmail.com>

fujitatomoya commented on source/How-To-Guides/Setup-ROS-2-with-VSCode-and-Docker-Container.rst:
```diff
@@ -176,7 +176,7 @@ Open the Dockerfile and add the following contents:
 
     FROM ros:ROS_DISTRO
-    ARG USERNAME=USERNAME
+    ARG USERNAME=YOUR_USERNAME
```
Do we really need to change here? i think `USERNAME` has been assigned as `build.args.USERNAME` in devcontainer.json? So i do not think we actually need to change this line?

tixel replied:
I checked the devcontainer docs
https://code.visualstudio.com/remote/advancedcontainers/add-nonroot-user

> While any images or Dockerfiles that come from the Dev Containers extension will include a non-root user with a UID/GID of 1000 (typically either called vscode or node), many base images and Dockerfiles do not. Fortunately, you can update or create a Dockerfile that adds a non-root user into your container.

They state that you need to add a specific non-root user if not present in the base image.

I checked the docker image for ros:iron and they do not create a non-root user:
https://hub.docker.com/layers/library/ros/iron/images/sha256-b674f8529dea243ddba377da120e121817f147de701e9b0e0fda8bde418486e0?context=repo&tab=vulnerabilities

So I do think it is necessary.

fujitatomoya replied:
hmm, is that true?

https://containers.dev/implementors/json_reference/

`build.args`: A set of name-value pairs containing [Docker image build arguments](https://docs.docker.com/engine/reference/commandline/build/#build-arg) that should be passed when building a Dockerfile. Environment and [pre-defined variables](https://containers.dev/implementors/json_reference/#variables-in-devcontainerjson) may be referenced in the values.

i tries this procedure with `devcontainer.json` (search under `tomoyafujita`). (I did not change the `Dockerfile`)

```json
{
    "name": "ROS 2 Development Container",
    "privileged": true,
    "remoteUser": "tomoyafujita",
    "build": {
        "dockerfile": "Dockerfile",
        "args": {
            "USERNAME": "tomoyafujita"
        }
    },
    "workspaceFolder": "/home/ws",
    "workspaceMount": "source=${localWorkspaceFolder},target=/home/ws/src,type=bind",
    "customizations": {
        "vscode": {
            "extensions":[
                "ms-vscode.cpptools",
                "ms-vscode.cpptools-themes",
                "twxs.cmake",
                "donjayamanne.python-extension-pack",
                "eamodio.gitlens",
                "ms-iot.vscode-ros"
            ]
        }
    },
    "containerEnv": {
        "DISPLAY": "unix:0",
        "ROS_AUTOMATIC_DISCOVERY_RANGE": "LOCALHOST",
        "ROS_DOMAIN_ID": "42"
    },
    "runArgs": [
        "--net=host",
        "-e", "DISPLAY=${env:DISPLAY}"
    ],
    "mounts": [
       "source=/tmp/.X11-unix,target=/tmp/.X11-unix,type=bind,consistency=cached",
        "source=/dev/dri,target=/dev/dri,type=bind,consistency=cached",
        "source=${localWorkspaceFolder}/../cache/rolling/build,target=/home/ws/build,type=bind",
        "source=${localWorkspaceFolder}/../cache/rolling/install,target=/home/ws/install,type=bind",
        "source=${localWorkspaceFolder}/../cache/rolling/log,target=/home/ws/log,type=bind"
    ],
    "postCreateCommand": "sudo rosdep update && sudo rosdep install --from-paths src --ignore-src -y && sudo chown -R $(whoami) /home/ws/"
}
```

then, container comes up with `tomoyafujita` user,

![image](https://github.com/ros2/ros2_documentation/assets/43395114/ecbea32c-19c1-470e-92f8-18e403f3bb69)

> I checked the docker image for ros:iron and they do not create a non-root user

that is expected, base images usually specify that since those are supposed to be the user configuration?

so after all, i still think we only need to apply your PR to `devcontainer.json` but `Dockerfile`. what do you think?

tixel replied:
I tested and it works now for me.
What I was doing wrong before was not putting my devcontainer.json in a ./devcontainer folder. In that case the passing of args from devcontainer to dockerfile break apparently.
So you are right: only the devcontainer.json needs changes.

fujitatomoya commented on source/How-To-Guides/Setup-ROS-2-with-VSCode-and-Docker-Container.rst:
```diff
@@ -201,7 +201,7 @@ Open the Dockerfile and add the following contents:
     CMD ["/bin/bash"]
 
-Search here also for the ``USERNAME`` and replace it with your ``Linux username`` and the ``ROS_DISTRO`` with the ROS 2 distribution you wish to use and added to the cache previously.
+Search here also for the ``YOUR_USERNAME`` and replace it with your ``Linux username`` and the ``ROS_DISTRO`` with the ROS 2 distribution you wish to use and added to the cache previously.
```
with the same reason, we should already have `USERNAME` set by user here?

```suggestion
Search here also for ``ROS_DISTRO`` with the ROS 2 distribution you wish to use and added to the cache previously.
```

tixel replied:
Following these instructions literally, did not work. I had to make these changes.

fujitatomoya commented on source/How-To-Guides/Setup-ROS-2-with-VSCode-and-Docker-Container.rst:
```diff
@@ -155,13 +155,13 @@ Therefore add the following to ``.devcontainer/devcontainer.json``:
             "source=${localWorkspaceFolder}/../cache/ROS_DISTRO/log,target=/home/ws/log,type=bind"
         ],
-        "postCreateCommand": "sudo rosdep update && sudo rosdep install --from-paths src --ignore-src -y && sudo chown -R USERNAME /home/ws/"
+        "postCreateCommand": "sudo rosdep update && sudo rosdep install --from-paths src --ignore-src -y && sudo chown -R YOUR_USERNAME /home/ws/"
```
```suggestion
        "postCreateCommand": "sudo rosdep update && sudo rosdep install --from-paths src --ignore-src -y && sudo chown -R $(whoami) /home/ws/"
```

tixel committed:
Apply suggestions from code review

I tested and it works now for me.
What I was doing wrong before was not putting my devcontainer.json in a ./devcontainer folder. In that case the passing of args from devcontainer to dockerfile break apparently.

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Tim <timbosschaerts@gmail.com>

tixel committed:
Update Setup-ROS-2-with-VSCode-and-Docker-Container.rst

revert Dockerfile to original.

Signed-off-by: Tim <timbosschaerts@gmail.com>

fujitatomoya reviewed (approved):
lgtm with green CI.

clalancette commented:
@Mergifyio backport rolling humble
